CRM/Financial/BAO/Payment.php

```diff
@@ -490,4 +486,18 @@ protected static function getContributionAndParamsInFormatForRecordFinancialTran
     return [$contributionDAO, $params];
   }
 
+  /**
+   * Does this payment complete the contribution
+   *
+   * @param int $contributionID
+   * @param float $paymentAmount
+   *
+   * @return bool
+   */
+  protected static function isPaymentCompletesContribution($contributionID, $paymentAmount) {
+    $outstandingBalance = CRM_Contribute_BAO_Contribution::getContributionBalance($contributionID);
```

Comment:
@monishdeb see here ^^ I'm getting the balance

Reply:
Yes agree, but my point was we are ONLY considering this comparison with balance before making payment. Not after as we need to ensure if the contribution is actually completed so that we can decide wether to call ```civicrm_api3('Contribution', 'completetransaction', $params)``` or not. Shouldn't we call this API ONLY when a contribution is completed?

Reply:
"Shouldn't we call this API ONLY when a contribution is completed?"

Yes - but we are comparing the amount being paid with the amount outstanding - so we should be able to determine if it will complete the contribution shouldn't we?

Reply:
Lets consider a use-case:
1. Create a contribution using price-fee $10
2. Then using line-item editor one changes the price fee to $30. It changes the Contribution to partially paid as $20 need to recorded
3. Then a ext call civicrm_api3('Payment', 'create', ['total_amount' => $10, ...]); 
NOTE: here we are not fully paying the contribution. 

Now as per the patch Payment.create fn, it will still call Contribtuion.completetransaction API as we are not rechecking the balance with contributions total amount after payment has been made. Am I missing something?

Reply:
@monishdeb so here we don't check AFTER payment is made but we check the balance immediately BEFORE any payment is made - the line item editor trxn should already have altered the balance so we should be able to calculate the balance immediately before applying the payment. We ALSO know the payment amount so we are checking before payment but we are checking if 

```
$outstandingBalance - $this PaymentAmount == 0;
```

Which should give the same amount as calculated if adding the payment changed it to 0

Reply:
Sorry, @eileenmcnaughton I see your point.

Reply:
So here ```getContributionBalance()``` get the remaining outstanding balance owed or refund. And compares with the total amount passed to the API. In that way we don;t need to check again wether to call completetransaction or not. **sigh** my throught process is getting rusty. Apologies :/

Reply:
lol - feel free to let it sit for a day or 2 & look again - it's crazy code to unravel!

Reply:
Hmm. But I am pretty confident with this patch, that it leaves all the stone unturned after this wind of code improvement and clarity.  But you never know what nasty stuff our test suite has to reveal after test build. Let's see what failures it has to spill.

Reply:
:-)